### PR TITLE
[tune] Make BayesOptSearch float-hash precision configurable

### LINKED ISSUE #50315: Ray Tune: `BayesOptSearch` config hashing unnecessarily skips suggested trials.
### What happened + What you expected to happen

Hi all,

I'm using Ray Tune's integration with the `BayesianOptimization` package via `BayesOptSearch` ([link to docs](https://docs.ray.io/en/latest/tune/examples/bayesopt_example.html)). To start, Ray Tune doesn't support the latest version of `BayesianOptimization` ([v2.0.3](https://github.com/bayesian-optimization/BayesianOptimization/releases/tag/v2.0.3)), so we have to downgrade to v1.5.1. My main issue, however, is how `BayesOptSearch` determines whether two sampled trials are duplicates. Specifically, after identifying a suggested `config`, we hash the dictionary so that we can check if we've already computed that value:

https://github.com/ray-project/ray/blob/6037e33c1ccd61dfd6335f160bb4423008ad840f/python/ray/tune/search/bayesopt/bayesopt_search.py#L275-L281

The problem here is that `_dict_hash` encodes floating point parameters to a certain `precision`, which `BayesOptSearch` sets to 5 by default:

https://github.com/ray-project/ray/blob/6037e33c1ccd61dfd6335f160bb4423008ad840f/python/ray/tune/search/bayesopt/bayesopt_search.py#L163-L164

https://github.com/ray-project/ray/blob/6037e33c1ccd61dfd6335f160bb4423008ad840f/python/ray/tune/search/bayesopt/bayesopt_search.py#L35

This seems like a very poor default, especially when dealing with floating point values. For example, if a user is only tuning a value like a neural network's learning rate, this precision would encode `1.46532e-05` and `1.00001e-05` as the same string: `"0.00001"`. As a user, that's pretty confusing. Instead, it seems prudent to either (1) increase the default precision to something more reasonable, like 8 or 16, or (2) allow the user to provide a precision (and call out the implications in the docs). It's fairly easy to manually set `repeat_float_precision = 16` after initialization, but it's just a bit annoying.

Let me know if you have any questions. Thank you!

### Versions / Dependencies

`ray[tune]==2.9.0`

### Reproduction script

(Sorry it's a bit longer than 50 lines; these methods are defined under private paths. I've added my own print statements.)

```python
import copy
import json
from typing import Dict

def flatten_dict(
    dt: Dict,
    delimiter: str = "/",
    prevent_delimiter: bool = False,
    flatten_list: bool = False,
):
    """Flatten dict.

    Output and input are of the same dict type.
    Input dict remains the same after the operation.
    """

    def _raise_delimiter_exception():
        raise ValueError(
            f"Found delimiter `{delimiter}` in key when trying to flatten "
            f"array. Please avoid using the delimiter in your specification."
        )

    dt = copy.copy(dt)
    if prevent_delimiter and any(delimiter in key for key in dt):
        # Raise if delimiter is any of the keys
        _raise_delimiter_exception()

    while_check = (dict, list) if flatten_list else dict

    while any(isinstance(v, while_check) for v in dt.values()):
        remove = []
        add = {}
        for key, value in dt.items():
            if isinstance(value, dict):
                for subkey, v in value.items():
                    if prevent_delimiter and delimiter in subkey:
                        # Raise if delimiter is in any of the subkeys
                        _raise_delimiter_exception()

                    add[delimiter.join([key, str(subkey)])] = v
                remove.append(key)
            elif flatten_list and isinstance(value, list):
                for i, v in enumerate(value):
                    if prevent_delimiter and delimiter in subkey:
                        # Raise if delimiter is in any of the subkeys
                        _raise_delimiter_exception()

                    add[delimiter.join([key, str(i)])] = v
                remove.append(key)

        dt.update(add)
        for k in remove:
            del dt[k]
    return dt


def _dict_hash(config, precision):
    print(f"{config=}; {precision=}")
    flatconfig = flatten_dict(config)
    print(f"{flatconfig=}")
    for param, value in flatconfig.items():
        if isinstance(value, float):
            flatconfig[param] = "{:.{digits}f}".format(value, digits=precision)

    print(f"{flatconfig=}")
    hashed = json.dumps(flatconfig, sort_keys=True, default=str)
    print(f"{hashed=}")
    return hashed


config_1 = {"learning_rate": 1.00001e-05}
config_2 = {"learning_rate": 1.46532e-05}
repeat_float_precision = 5

print(f"=== Config Hash 1 ===")
config_hash_1 = _dict_hash(config_1, repeat_float_precision)
print(f"=== Config Hash 2 ===")
config_hash_2 = _dict_hash(config_2, repeat_float_precision)
assert config_hash_1 == config_hash_2
```

### Issue Severity

Low: It annoys or frustrates me.

### PR DESCRIPTION
**Pull request by saivedant169**
## Why are these changes needed?

`BayesOptSearch` hashes each suggested config to decide whether it has already been tried, and `_dict_hash` encodes float values at a fixed precision of 5 decimals (`repeat_float_precision = 5`). For small-magnitude parameters that is too coarse. In a learning-rate search, `1.00001e-05` and `1.46532e-05` both round to `"0.00001"`, so two genuinely different suggestions look identical and one gets counted as a repeat and skipped. With `patience` set, enough false repeats can end the run early.

This exposes `repeat_float_precision` as a constructor argument so it can be raised when needed (the issue notes it can already be set by hand after init, just awkwardly). The default stays at 5 so existing runs are unaffected, and negative values are rejected.

Verified locally against the installed package: at precision 5 the two learning rates above hash to the same string, at precision 16 they stay distinct, and the default remains 5.

## Related issue number

Closes #50315

## Checks

- [x] Added tests in `test_searchers.py` (`BayesOptHashPrecisionTest`): hashing at low vs high precision, the new argument, and the negative-value guard. All pass.
- [x] `ruff check` passes on the changed files.

### CODE REVIEW

**Comment by saivedant169** [comment]
@gemini-code-assist check and let me know if there are any changes required to done.

File: python/ray/tune/search/bayesopt/bayesopt_search.py [hunk 1]
```diff
@@ -160,8 +165,11 @@ def __init__(
             assert mode in ["min", "max"], "`mode` must be 'min' or 'max'."
         self._config_counter = defaultdict(int)
         self._patience = patience
-        # int: Precision at which to hash values.
-        self.repeat_float_precision = 5
+        if repeat_float_precision < 0:
+            raise ValueError("repeat_float_precision must be non-negative.")
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply python/ray/tune/search/bayesopt/bayesopt_search.py:173]: Could you add this change

File: python/ray/tune/tests/test_searchers.py [hunk 2]
```diff
@@ -801,6 +801,41 @@ def testOptunaWithStorage(self):
...
+        from ray.tune.search.bayesopt import BayesOptSearch
+
+        with self.assertRaises(ValueError):
+            BayesOptSearch(repeat_float_precision=-1)
+
+        with self.assertRaises(TypeError):
+            BayesOptSearch(repeat_float_precision="5")
+
+        with self.assertRaises(TypeError):
+            BayesOptSearch(repeat_float_precision=5.5)
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply python/ray/tune/tests/test_searchers.py:836]: Add

**Review by pseudo-rnd-thoughts — CHANGES_REQUESTED** [review]
Thanks for the PR @saivedant169
Overall it looks good but could you implement the two Gemini suggested changes

**Comment by saivedant169** [comment]
@pseudo-rnd-thoughts both of those are already in, they landed in the last two commits rather than through the suggestion button, so the Gemini threads still show as pending. The constructor raises TypeError for non-int values and also rejects bool, since bool passes isinstance int. testInvalidRepeatFloatPrecisionRaises covers "5", 5.5 and True alongside the -1 ValueError case. Happy to change anything else.

**Comment by saivedant169** [comment]
@gemini-code-assist check and let me know if there are any changes required to be done.

File: python/ray/tune/search/bayesopt/bayesopt_search.py [hunk 3]
```diff
@@ -160,8 +165,13 @@ def __init__(
             assert mode in ["min", "max"], "`mode` must be 'min' or 'max'."
         self._config_counter = defaultdict(int)
         self._patience = patience
-        # int: Precision at which to hash values.
-        self.repeat_float_precision = 5
+        if not isinstance(repeat_float_precision, int):
+            raise TypeError("repeat_float_precision must be an integer.")
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply python/ray/tune/search/bayesopt/bayesopt_search.py]: Ahh I hadn't seen this comment so I didn't understand the bool and int check

**Review by pseudo-rnd-thoughts — APPROVED** [review]
LGTM
